src/tiled/wangfiller.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+ * wangfiller.cpp
+ * Copyright 2017, Benjamin Trotter <bdtrotte@ucsc.edu>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "wangfiller.h"
+
+#include "tilelayer.h"
+#include "wangset.h"
+
+using namespace Tiled;
+using namespace Internal;
+
+static const QPoint adjacentPoints[] = {
+    QPoint( 0, -1),
+    QPoint( 1, -1),
+    QPoint( 1,  0),
+    QPoint( 1,  1),
+    QPoint( 0,  1),
+    QPoint(-1,  1),
+    QPoint(-1,  0),
+    QPoint(-1, -1)
+};
+
+WangFiller::WangFiller(WangSet *wangSet)
+    :mWangSet(wangSet)
+{
+}
+
+void WangFiller::setWangSet(WangSet *wangSet)
+{
+    mWangSet = wangSet;
+}
+
+Cell WangFiller::findFittingCell(const TileLayer &back,
+                                 const TileLayer &front,
+                                 const QRegion &fillRegion,
+                                 QPoint point,
+                                 bool lookForward) const
```

Comment:
For both `findFittingCell` and `fillRegion`, the `lookForward` boolean is based on `!mWangFiller->wangSet()->isComplete()`. Maybe just remove the parameter and check completeness locally?